ocaml/xapi/xapi_host.ml

```diff
@@ -2904,14 +2905,97 @@ let apply_updates ~__context ~self ~hash =
   in
   Db.Host.set_last_software_update ~__context ~self
     ~value:(get_servertime ~__context ~host:self) ;
-  Repository.apply_immediate_guidances ~__context ~host:self ~guidances ;
   (* The warnings may not be returned to async caller if this happens
    * on the coordinator host and the 'restartToolstack' is required.
    *)
-  warnings
+  List.map
+    (fun g ->
+      [
+        Api_errors.updates_require_recommended_guidance
+      ; Updateinfo.Guidance.to_string g
+      ]
+    )
+    guidances
+  @ warnings
 
 let set_https_only ~__context ~self ~value =
   let state = match value with true -> "close" | false -> "open" in
   ignore
   @@ Helpers.call_script !Xapi_globs.firewall_port_config_script [state; "80"] ;
   Db.Host.set_https_only ~__context ~self ~value
+
+let restart_device_models_for_recommended_guidances ~__context ~host =
+  (* Restart device models of all running HVM VMs on the host by doing
+   * local migrations if it is required by recommended guidances. *)
+  Repository_helpers.do_with_device_models ~__context ~host
+    ~only_vm_with_recommended_guidances:true
+  @@ fun (ref, record) ->
+  match
+    (record.API.vM_power_state, Helpers.has_qemu_currently ~__context ~self:ref)
+  with
+  | `Running, true ->
+      Xapi_vm_migrate.pool_migrate ~__context ~vm:ref ~host
+        ~options:[("live", "true")] ;
+      None
+  | `Paused, true ->
+      error "VM 'ref=%s' is paused, can't restart device models for it"
+        (Ref.string_of ref) ;
+      Some ref
+  | _ ->
+      (* No device models are running for this VM *)
+      None
+
+let apply_recommended_guidances ~__context ~self =
```

Comment:
Another concern is that this function will run on a member host directly. I don't have strong argument for not doing this given now XAPI doesn't have special logic in handling, i.e. `host.reboot, in `message_forwarding` in coordinator. I think an alternative is just to handle this in coordinator, and send primitive XAPI calls, like `host.reboot` to member hosts.